source/level_set_okz_compute_normal.cc

```diff
@@ -0,0 +1,261 @@
+
+      normal_values.integrate(true, false);
+      normal_values.distribute_local_to_global(dst);
+    }
+}
+
+
+
+template <int dim>
+void
+LevelSetOKZSolverComputeNormal<dim>::compute_normal_vmult(
+  BlockVectorType &      dst,
+  const BlockVectorType &src) const
+{
+  dst = 0.;
+#define OPERATION(c_degree, dummy)                                                     \
+  this->matrix_free.cell_loop(&LevelSetOKZSolverComputeNormal<                         \
+                                dim>::template local_compute_normal<c_degree, double>, \
+                              this,                                                    \
+                              dst,                                                     \
+                              src)
+
+  EXPAND_OPERATIONS(OPERATION);
+#undef OPERATION
+
+  // The number "4" below is so that constraints_normals is used
+  for (unsigned int i = 0; i < this->matrix_free.get_constrained_dofs(4).size(); ++i)
+    for (unsigned int d = 0; d < dim; ++d)
+      dst.block(d).local_element(this->matrix_free.get_constrained_dofs(4)[i]) =
+        preconditioner.get_vector().local_element(
+          this->matrix_free.get_constrained_dofs(4)[i]) *
+        src.block(d).local_element(this->matrix_free.get_constrained_dofs(4)[i]);
+}
+
+
+
+template <int dim>
+struct ComputeNormalMatrix
+{
+  ComputeNormalMatrix(const LevelSetOKZSolverComputeNormal<dim> &problem)
+    : problem(problem)
+  {}
+
+  template <typename Number>
+  void
+  vmult(LinearAlgebra::distributed::BlockVector<Number> &      dst,
+        const LinearAlgebra::distributed::BlockVector<Number> &src) const
+  {
+    problem.compute_normal_vmult(dst, src);
+  }
+
+  const LevelSetOKZSolverComputeNormal<dim> &problem;
+};
+
+
+
+// @sect4{LevelSetOKZSolverComputeNormal::compute_normal}
+template <int dim>
+void
+LevelSetOKZSolverComputeNormal<dim>::compute_normal(const bool fast_computation)
+{
+  // This function computes the normal from a projection of $\nabla C$ onto
+  // the space of linear finite elements (with some small damping)
+
+  // compute right hand side
+  this->normal_vector_rhs = 0;
+#define OPERATION(c_degree, u_degree)                                                  \
+  this->matrix_free.cell_loop(                                                         \
+    &LevelSetOKZSolverComputeNormal<dim>::template local_compute_normal_rhs<c_degree>, \
+    this,                                                                              \
+    this->normal_vector_rhs,                                                           \
+    this->vel_solution)
+
+  EXPAND_OPERATIONS(OPERATION);
+#undef OPERATION
+
+  if (this->parameters.approximate_projections == true)
+    {
+      AssertThrow(false, ExcNotImplemented());
+      // [PM]: The following code has been removed since it was not used in any test
+      // and it introduced annoying cyclic dependencies.
```

Comment:
This is experimental code anyway, so it can remain in this state until there appears a need for it again with some additional experiments.